proof-systems/src/darlin/tests/final_darlin.rs

```diff
@@ -0,0 +1,353 @@
+            |lc| lc + d,
+        );
+
+        Ok(())
+    }
+}
+
+impl<G1, G2> PCDCircuit<G1> for TestCircuit<G1, G2>
+    where
+        G1: AffineCurve<BaseField = <G2 as AffineCurve>::ScalarField> + ToConstraintField<<G2 as AffineCurve>::ScalarField>,
+        G2: AffineCurve<BaseField = <G1 as AffineCurve>::ScalarField> + ToConstraintField<<G1 as AffineCurve>::ScalarField>,
+{
+    type SetupData       =  CircuitInfo<G1, G2>;
+    type IncrementalData =  (G1::ScalarField, G1::ScalarField);
+    type SystemInputs    =  FinalDarlinDeferredData<G1, G2>;
+    type PreviousPCD     =  TestPrevPCD<G1, G2>;
+
+    fn init(config: Self::SetupData) -> Self {
+        Self {
+            a: None,
+            b: None,
+            c: None,
+            d: None,
+            num_constraints: config.num_constraints,
+            num_variables: config.num_variables,
+            deferred: config.dummy_deferred.clone()
+        }
+    }
+
+    fn init_state(
+        config:               Self::SetupData,
+        previous_proofs_data: Vec<Self::PreviousPCD>,
+        _previous_proofs_vks: Vec<<Self::PreviousPCD as PCD>::PCDVerifierKey>,
+        incremental_data:     Self::IncrementalData
+    ) -> Self
+    {
+        assert_eq!(previous_proofs_data.len(), 1);
+
+        let a = incremental_data.0;
+        let b = incremental_data.1;
+
+        let mut c = a;
+        c.mul_assign(&b);
+        let mut d = c;
+        d.mul_assign(&b);
+
+        Self {
+            a: Some(a),
+            b: Some(b),
+            c: Some(c),
+            d: Some(d),
+            num_constraints: config.num_constraints,
+            num_variables: config.num_variables,
+            deferred: previous_proofs_data[0].0.clone(),
+        }
+    }
+
+    fn get_sys_ins(&self) -> Result<&Self::SystemInputs, PCDError> {
+        Ok(&self.deferred)
+    }
+
+    fn get_usr_ins(&self) -> Result<Vec<G1::ScalarField>, PCDError> {
+        let c = self.c.ok_or(PCDError::MissingUserInputs("c".to_owned()))?;
+        let d = self.d.ok_or(PCDError::MissingUserInputs("d".to_owned()))?;
+        Ok(vec![c, d])
+    }
+}
+
+/// Generates a FinalDarlinPCD from TestCircuit1, given an instance of 
+/// FinalDarlinDeferred as previous PCD (via CircuitInfo). 
+/// The incremental data a,b is sampled randomly.
+#[allow(dead_code)]
+pub fn generate_test_pcd<'a, G1: AffineCurve, G2:AffineCurve, D: Digest + 'a, R: RngCore>(
+    pc_ck_g1: &CommitterKey<G1>,
+    final_darlin_pk: &FinalDarlinProverKey<G1::ScalarField, InnerProductArgPC<G1, D>>,
+    info: CircuitInfo<G1, G2>,
+    zk: bool,
+    rng: &mut R,
+) -> FinalDarlinPCD<'a, G1, G2, D>
+    where
+        G1: AffineCurve<BaseField = <G2 as AffineCurve>::ScalarField> + ToConstraintField<<G2 as AffineCurve>::ScalarField>,
+        G2: AffineCurve<BaseField = <G1 as AffineCurve>::ScalarField> + ToConstraintField<<G1 as AffineCurve>::ScalarField>,
+{
+    let prev_pcds = vec![TestPrevPCD::<G1, G2>(info.dummy_deferred.clone())];
```

Comment:
Why are the ``prev_pcds`` taken from ``CircuitInfo`` instead of ``TestCircuit``?

Reply:
It's just for testing: since I already generated one for the CircuitInfo I'm just re-using it instead of generating a new one

Reply:
Although it is just for testing, it'd be a better hint on the different purposes of ``Circuit Info`` and ``TestCircuit``, no?